### PR TITLE
Consistent LogicalPlan subquery handling in TreeNode::apply and TreeNode::visit 

### PR DESCRIPTION
**Pull request by peter-toth**
## Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/issues/8913.

## Rationale for this change

`TreeNode` APIs are inconsistent on `LogicaPlan` because `LogicalPlan:apply()` and `LogicalPlan::visit()` are recursing into subquery plans too (these methods override the original `TreeNode::apply` and `TreeNode::visit()` methods).
But there is no `LogicalPlan:rewite()` or `LogicalPlan::transform()` overrides defined so calling `rewrite()` and `transform()` on a `LogicalPlan` does rewrite/transform only the current level of the plan and doesn't touch subquery plans.

## What changes are included in this PR?

This PR:
- Moves the code of `LogicalPlan::apply()` to `LogicalPlan::apply_with_subqueries()`, and `LogicalPlan::visit()` to `LogicalPlan::visit_with_subqueries()` to separate functionality from original the `TreeNode::apply()` and `TreeNode::visit()`. So from now the new explicit `LogicalPlan::..._with_subqueries()` APIs do traverse into subquery plans, while the original `TreeNode` APIs don't.
- Adds `LogicalPlan::rewrite_with_subqueries()`, `LogicalPlan::transform_with_subqueries()`,  `LogicalPlan::transform_down_with_subqueries()`, `LogicalPlan::transform_down_mut_with_subqueries()`, `LogicalPlan::transform_up_with_subqueries()`, `LogicalPlan::transform_up_mut_with_subqueries()` and `LogicalPlan::transform_down_up_with_subqueries()` to be able to rewrite `LogicalPlan`s with their subquery plans.
- Fixes the issue that subquery plans were not treated as child of the node. In the `LogicalPlan::..._with_subqueries()`s APIs the subquery plans behave like the first children of a node (siblings of the real children of the node).

## Are these changes tested?

Yes, with Existing UTs.

## Are there any user-facing changes?

Yes -  `LogicalPlan::apply()` no longer recurses into subqueries. The API user needs to change the code to `LogicalPlan::apply_with_subqueries()` to get the old behaviour.

### CODE REVIEW

File: datafusion/optimizer/src/analyzer/mod.rs [hunk 1]
```diff
@@ -155,7 +155,7 @@ impl Analyzer {
 
 /// Do necessary check and fail the invalid plan
 fn check_plan(plan: &LogicalPlan) -> Result<()> {
-    plan.apply(&mut |plan: &LogicalPlan| {
+    plan.apply_with_subqueries(&mut |plan: &LogicalPlan| {
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/mod.rs:158]: It is interesting that this didn't change any plans / tests.

**Review by alamb — COMMENTED** [review]
Thank you (again) @peter-toth  -- this looks really nice as well

I think there are is some more documentation we could add, but overall the only concern I have about this PR is the lack of test changes (like it seems to change the semantics of the analyzer / correlated expressions but no tests are changed). So this perhaps suggests we have a lack of test coverage in this area :thin

File: datafusion/optimizer/src/analyzer/mod.rs (hunk 1 above)
Comment (peter-toth) [reply datafusion/optimizer/src/analyzer/mod.rs:158]: Basically I just moved the code of `LogicalPlan:: apply()` (that overrided `TreeNode:: apply()`) to `LogicalPlan:: apply_with_subqueries()` and changed the `LogicalPlan:: apply()` calls to `LogicalPlan:: apply_with_subqueries()`. I updated the PR description to be clear.

File: datafusion/expr/src/logical_plan/plan.rs [hunk 2]
```diff
@@ -1080,8 +1298,72 @@ impl LogicalPlan {
...
+            TreeNodeRecursion::Continue => {
+                self.visit_subqueries(visitor)?;
+                handle_visit_recursion!(
+                    self.apply_children(&mut |n| n.visit_with_subqueries(visitor))?,
+                    UP
+                );
+                visitor.f_up(self)
+            }
+            TreeNodeRecursion::Jump => {
+                self.visit_subqueries(visitor)?;
```
[the remark this answers is not in the document]
Comment (peter-toth) [reply datafusion/expr/src/logical_plan/plan.rs:1328]: I just moved the code here from `LogicalPlan::visit()` and in this PR I didn't want to change how subqueries are handled. The new code I added in `LogicalPlan::rewrite_with_subqueries()` are in sync with this behaviour.

But this is a good question and I think you are right and I too would treat them as additional children. But that would be an API behavior change so I would rather change that in a separate PR.

**Comment by peter-toth** [comment]
> Thank you (again) @peter-toth -- this looks really nice as well
> 
> I think there are is some more documentation we could add, but overall the only concern I have about this PR is the lack of test changes (like it seems to change the semantics of the analyzer / correlated expressions but no tests are changed). So this perhaps suggests we have a lack of test coverage in this area 🤔

I rephrased the PR description to be clear. This PR doesn't change the analyzer or anything that uses `LogicalPlan` APIs. This PR just moves the overriden `TreeNode` APIs (`apply()`/`visit()`) in `LogicalPlan` to under new names so as to be clear they recurse into subqueries. Also, this PR adds the missing `LogicalPlan::rewrite_with_subqueries()` that is capable to rewrite `LogicalPlan`s with their subqueries.

**Comment by peter-toth** [comment]
Let me update this PR with the latest `main` (including https://github.com/apache/arrow-datafusion/pull/9876) and address review feedbacks today. Until then I'm marking this PR as draft.

**Comment by peter-toth** [comment]
@alamb , while I was working on this I realized that currently `TreeNode` recursion definitions are a bit hard to follow and could be simplified, which will be useful this PR as well. I've opened https://github.com/apache/arrow-datafusion/pull/9965 to do that. Can we please review/merge that https://github.com/apache/arrow-datafusion/pull/9965 first, then I can wrap up this PR.

**Comment by peter-toth** [comment]
@alamb , I've updated the PR. I will try to add tests for the `LogicalPlan::..._with_subqueries()` APIs in this PR or a in follow-up PR on Monday.

**Comment by alamb** [comment]
> @alamb, I've updated the PR. I will try to add tests for the `LogicalPlan::..._with_subqueries()` APIs in this PR or a in follow-up PR on Monday.

I think it would be great to do this as a follow up PR on Monday, however I think the utility of unit tests on this code is somewhat limited. It will be tested end to end once I switch the optimizers to using the TreeNode API in https://github.com/apache/arrow-datafusion/pull/9954. 

I took the liberty of pushing a commit to this branch to fix clippy and plan to merge it in when CI passes

File: datafusion/expr/src/logical_plan/plan.rs [hunk 3]
```diff
@@ -285,57 +287,57 @@ impl LogicalPlan {
     /// Calls `f` on all expressions (non-recursively) in the current
     /// logical plan node. This does not include expressions in any
     /// children.
-    pub fn inspect_expressions<F, E>(self: &LogicalPlan, mut f: F) -> Result<(), E>
-    where
-        F: FnMut(&Expr) -> Result<(), E>,
-    {
+    pub fn apply_expressions<F: FnMut(&Expr) -> Result<TreeNodeRecursion>>(
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/plan.rs:290]: This is an api change as it renames a public function. I couldn't find a way to get the error type to line up, so I think it is unavoidable.

I'll mark this PR as API change

**Review by alamb — DISMISSED** [review]
Thank you @peter-toth . Love it

**Comment by alamb** [comment]
🤔  it appears this PR has some sort of infinite loop -- I will investigate

**Comment by peter-toth** [comment]
> 🤔  it appears this PR has some sort of infinite loop -- I will investigate

Thanks @alamb . Maybe because now we treat subqueries as additional children? I can also look into it later today or tomorrow...

**Comment by alamb** [comment]
> > 🤔  it appears this PR has some sort of infinite loop -- I will investigate
> 
> Thanks @alamb . Maybe because now we treat subqueries as additional children? I can also look into it later today or tomorrow...

Thanks @peter-toth  -- yes I think you are correct. I debugged a bit and I found that calling apply_subqueries cased infinite recursion.

I removed the apply_subqueries call (see this https://github.com/peter-toth/arrow-datafusion/pull/3) but then some of the subquery plans don't work

So now I am confused -- should `apply` visit all subquery `LogicalPlan`s or not? Maybe it does make sense, as you said above, to have separate `apply` and `apply_subquery` calls 🤔

File: datafusion/expr/src/logical_plan/plan.rs (hunk 3 above)
Comment (alamb) [reply datafusion/expr/src/logical_plan/plan.rs:290]: Actually, I found another pattern here: https://github.com/apache/arrow-datafusion/blob/d201ec742b08fa2a0549fc27f0b39622d67391ab/datafusion/expr/src/utils.rs#L684

Maybe we can pull this part (add `apply_expressions` into a separate PR)? I think it would be valuable in its own -- for example it could be used to void the `expressions()` copy here:

https://github.com/apache/arrow-datafusion/blob/27bdf3c8c976874aae2417cb2aa435448832ecc0/datafusion/optimizer/src/analyzer/subquery.rs#L236-L241

It could also make this PR smaller

Comment (peter-toth) [reply datafusion/expr/src/logical_plan/plan.rs:290]: Yes, we can do that. Let me file a new PR with this change tomorrow and then look into the infinite recursion issue of this PR (https://github.com/peter-toth/arrow-datafusion/pull/3). Until then I'm marking this PR as draft.

**Comment by peter-toth** [comment]
This is now ready for review again. The infinite recursion was due to an extra accidental. `apply()` call: https://github.com/apache/arrow-datafusion/pull/9913/commits/c72349217e0dde13d972e71fa99ae4e5e6025cdf

File: datafusion/expr/src/logical_plan/plan.rs (hunk 3 above)
Comment (peter-toth) [reply datafusion/expr/src/logical_plan/plan.rs:290]: I've opened https://github.com/apache/arrow-datafusion/pull/9984 as I think we don't need that `inspect_expr_pre()`, this PR together with https://github.com/apache/arrow-datafusion/pull/9984 will remove all usecases of it.

Comment (peter-toth) [reply datafusion/expr/src/logical_plan/plan.rs:290]: As for https://github.com/apache/arrow-datafusion/blob/27bdf3c8c976874aae2417cb2aa435448832ecc0/datafusion/optimizer/src/analyzer/subquery.rs#L236-L241 I plan to check all `.expressions()` calls and see if it can be transformed to the better `.apply_expressions()` once this PR is merged.

Comment (alamb) [reply datafusion/expr/src/logical_plan/plan.rs:311]: Sounds good

Comment (alamb) [reply datafusion/expr/src/logical_plan/plan.rs:311]: > I've opened #9984 as I think we don't need that `inspect_expr_pre()`. This PR together with #9984 will remove all usecases of it.

I merged #9984

In order to make this PR easier to merge (so it isn't an API change) I 
1. pushed 7c90e70dc to restore `inspect_expressions` (by calling `apply_expressions`) and deprecated it. 
2. Removed the API change

**Review by alamb — APPROVED** [review]
Thank you @peter-toth  -- this PR looks great to me. 

I took the liberty of restring `inspect_expressions` and marking it deprecated and merging up from main.

As I have several PRs backed up on this one, I would like to merge it asap so plan to do so first thing tomorrow morning unless anyone else would like time to review

File: datafusion/expr/src/logical_plan/plan.rs (hunk 3 above)
Comment (alamb) [reply datafusion/expr/src/logical_plan/plan.rs:311]: > I plan to check all .expressions() calls and see if it can be transformed to the better .apply_expressions() once this PR is merged.

This would be good -- I think there are some more (though https://github.com/apache/arrow-datafusion/pull/9974 did get merged)

**Comment by alamb** [comment]
I am not sure what you are planning on working on next @peter-toth but one thing I noticed is that keeping track of all the `LogicalPlan` tree node walking / rewriting is somewhat hard for me (and `datafusion/expr/src/logical_plan/plan.rs` is getting pretty large).


 I wonder if we should move all the code into a single module (maybe alongside the `impl TreeNode for LogicalPlan`) ?  Maybe somethign like  `datafusion/expr/src/logical_plan/tree_node.rs` or `datafusion/expr/src/logical_plan/apply_visit.rs` 🤔

File: datafusion/expr/src/logical_plan/plan.rs [hunk 4]
```diff
@@ -1079,57 +1317,178 @@ impl LogicalPlan {
...
-                    | Expr::InSubquery(InSubquery { subquery, .. })
-                    | Expr::ScalarSubquery(subquery) => {
-                        // use a synthetic plan so the collector sees a
-                        // LogicalPlan::Subquery (even though it is
-                        // actually a Subquery alias)
-                        let synthetic_plan = LogicalPlan::Subquery(subquery.clone());
-                        synthetic_plan.apply(op)?;
-                    }
-                    _ => {}
+    pub fn visit_with_subqueries<V: TreeNodeVisitor<Node = Self>>(
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/plan.rs:1349]: as a follow on PR I also plan to add documentation to these functions. Perhaps I can do so as part of moving the code into its own module

**Comment by peter-toth** [comment]
> I took the liberty of restring `inspect_expressions` and marking it deprecated and merging up from main.

Sure, thanks for doing that!

> Removed the API change

I was thinking about this a bit more and I unfortunately think this PR is still an API change. If some external code calls `.apply()` on a `LogicalPlan` then this PR might break it. This is because the implementation of `.apply()`will be de default implementation of `TreeNode::apply()` after this PR and no longer recurse into subqueries. The API user needs to change the code to `.apply_with_subqueries()` to get the old behaviour.
But I have no other idea to fix the current inconsistency between `.apply()`/`.visit()`, that do recurse into subqueries and `.rewrite()`/`.transform()`, that don't. (Well we could change `.rewrite()`/`.transform()` to do recurse into subqueries, but that too would be a breaking change.)

> I am not sure what you are planning on working on next @peter-toth but one thing I noticed is that keeping track of all the LogicalPlan tree node walking / rewriting is somewhat hard for me (and datafusion/expr/src/logical_plan/plan.rs is getting pretty large).

I agree. I'm happy to take that task, but I will be travelling from Tuesday till the end of the week, so most likely I can do that once I'm back.

**Comment by alamb** [comment]
> I was thinking about this a bit more and I unfortunately think this PR is still an API change.

That makes sense -- I have added the label back.

> But I have no other idea to fix the current inconsistency between 

I think it is inevitable that there are different calls for `LogicalPlan` for traversing with subqueries and without subqueries


> I agree. I'm happy to take that task,

Awesome -- thank you. I filed https://github.com/apache/arrow-datafusion/issues/9994 to track the idea. I may make a few small PRs in this direction to start the process too

**Comment by alamb** [comment]
Thanks again @peter-toth

File: datafusion/expr/src/logical_plan/plan.rs [hunk 5]
```diff
@@ -1079,57 +1317,178 @@ impl LogicalPlan {
     }
 }
 
+/// This macro is used to determine continuation during combined transforming
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/plan.rs:1320]: I think this may be dead code -- can somone check https://github.com/apache/arrow-datafusion/pull/9997 ?

File: datafusion/expr/src/logical_plan/plan.rs (hunk 4 above)
Comment (alamb) [reply datafusion/expr/src/logical_plan/plan.rs:1349]: Docs: https://github.com/apache/arrow-datafusion/pull/9996

**Comment by peter-toth** [comment]
Thanks @alamb for the review!

File: datafusion/expr/src/logical_plan/plan.rs (hunk 5 above)
Comment (peter-toth) [reply datafusion/expr/src/logical_plan/plan.rs:1320]: Actually, I deliberately added `..._with_subqueries()` versions of all base `TreeNode` APIs to `LogicalPlan`. These 3 macros are used in `LogicalPlan::rewrite_with_subqueries()`, `LogicalPlan::transform_down_with_subqueries()` and `LogicalPlan::transform_up_with_subqueries()`.

Comment (alamb) [reply datafusion/expr/src/logical_plan/plan.rs:1320]: Cool -- thanks -- I think that will be clearer as we consolidate the tree rewriting code
